Mathlib/MeasureTheory/Measure/WithDensity.lean

```diff
@@ -269,51 +270,90 @@ theorem withDensity_apply_eq_zero {f : α → ℝ≥0∞} {s : Set α} (hf : Mea
+             fun h ↦ by rw [h] at ha; exact ha⟩
+    rw [measure_congr t, withDensity_congr_ae hfg]
+    have M : MeasurableSet { x : α | g x = 0 } := hg (measurableSet_singleton _)
+    rw [withDensity_apply _ M, lintegral_eq_zero_iff hg]
+    filter_upwards [ae_restrict_mem M]
+    simp only [imp_self, Pi.zero_apply, imp_true_iff]
+
+theorem withDensity_apply_eq_zero {f : α → ℝ≥0∞} {s : Set α} (hf : Measurable f) :
+    μ.withDensity f s = 0 ↔ μ ({ x | f x ≠ 0 } ∩ s) = 0 :=
+  withDensity_apply_eq_zero' <| Measurable.aemeasurable hf
 #align measure_theory.with_density_apply_eq_zero MeasureTheory.withDensity_apply_eq_zero
 
-theorem ae_withDensity_iff {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+theorem ae_withDensity_iff' {p : α → Prop} {f : α → ℝ≥0∞} (hf : AEMeasurable f μ) :
     (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ, f x ≠ 0 → p x := by
-  rw [ae_iff, ae_iff, withDensity_apply_eq_zero hf, iff_iff_eq]
+  rw [ae_iff, ae_iff, withDensity_apply_eq_zero' hf, iff_iff_eq]
   congr
   ext x
   simp only [exists_prop, mem_inter_iff, iff_self_iff, mem_setOf_eq, not_forall]
-#align measure_theory.ae_with_density_iff MeasureTheory.ae_withDensity_iff
 
-theorem ae_withDensity_iff_ae_restrict {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+theorem ae_withDensity_iff {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+    (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ, f x ≠ 0 → p x :=
+  ae_withDensity_iff' <| Measurable.aemeasurable hf
+
+theorem ae_withDensity_iff_ae_restrict' {p : α → Prop} {f : α → ℝ≥0∞}
+    (hf : AEMeasurable f μ) :
     (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ.restrict { x | f x ≠ 0 }, p x := by
-  rw [ae_withDensity_iff hf, ae_restrict_iff']
+  rw [ae_withDensity_iff' hf, ae_restrict_iff'₀]
   · simp only [mem_setOf]
-  · exact hf (measurableSet_singleton 0).compl
+  · rcases hf with ⟨g, hg, hfg⟩
+    have nonneg_eq_ae : {x | g x ≠ 0} =ᵐ[μ] {x | f x ≠ 0} := by {
+      filter_upwards [hfg] with a ha
+      simp only [eq_iff_iff]
+      exact ⟨fun (h : g a ≠ 0) ↦ by rwa [← ha] at h,
+             fun (h : f a ≠ 0) ↦ by rwa [ha] at h⟩
+    }
+    exact NullMeasurableSet.congr
+      (MeasurableSet.nullMeasurableSet
+        <| hg (measurableSet_singleton _)).compl
+      nonneg_eq_ae
+
+theorem ae_withDensity_iff_ae_restrict {p : α → Prop} {f : α → ℝ≥0∞} (hf : Measurable f) :
+    (∀ᵐ x ∂μ.withDensity f, p x) ↔ ∀ᵐ x ∂μ.restrict { x | f x ≠ 0 }, p x :=
+  ae_withDensity_iff_ae_restrict' <| Measurable.aemeasurable hf
 #align measure_theory.ae_with_density_iff_ae_restrict MeasureTheory.ae_withDensity_iff_ae_restrict
 
-theorem aemeasurable_withDensity_ennreal_iff {f : α → ℝ≥0} (hf : Measurable f) {g : α → ℝ≥0∞} :
+theorem aemeasurable_withDensity_ennreal_iff' {f : α → ℝ≥0}
+    (hf : AEMeasurable f μ) {g : α → ℝ≥0∞} :
     AEMeasurable g (μ.withDensity fun x => (f x : ℝ≥0∞)) ↔
       AEMeasurable (fun x => (f x : ℝ≥0∞) * g x) μ := by
+  have t : ∃f', Measurable f' ∧ f =ᵐ[μ] f' := hf; rcases t with ⟨f', hf'⟩
```

Comment:
Please split that line into two lines (we write only one tactic per line).
Also if you want to use `rcases` right away you could do it in one line with `obtain ⟨f', hf'⟩ := hf`. And you should probably split hf' further into two pieces (one for measurability, one for the a.e. equality) like you do elsewhere.

Reply:
The issue is that I want in the context `⟨f', ⟨hf'_m, hf'_ae⟩⟩` but I also want to keep `AEMeasurable f` written as is, to use `AEMeasurable.coe_nnreal_ennreal` easely. I could'nt find a tactic to like `rcases`or `obtain` that does not destruct the initial hyp. Do you have an idea?

Reply:
One option is the one I mentioned in my other comment: don't use `rcases` or `obtain` but refer to the measurable function with `hf.mk f` and to its properties with `hf.measurable_mk` and `hf.ae_eq_mk`.

Otherwise I agree that you have to copy the hypothesis if you want to both keep it and get its components.

Reply:
I tried to do that and, on some cases, it does not typecheck without help and it makes the notation quite heavy. I guess I'll keep what I've originally done and split in into two lines, even if it's not very elegant.